src/System.Private.ServiceModel/src/System/ServiceModel/Channels/ReadAheadWrappingStream.cs

```diff
@@ -0,0 +1,304 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+// 
+
+using System.Diagnostics.Contracts;
+using System.IO;
+using System.Runtime;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.ServiceModel.Channels
+{
+    internal class ReadAheadWrappingStream : DelegatingStream
+    {
+        public const string ReadAheadWrappingStreamPropertyName = "ReadAheadWrappingStreamProperty";
+        private readonly int _readAheadBytes;
+        private readonly ArraySegment<byte>[] _buffers;
+        private Task<int> _previousCompletedReadTask;
+        private Task<int> _previousCompletedEnsureBufferedTask;
+        private bool _eof = false;
+
+        public ReadAheadWrappingStream(Stream innerStream, int readAheadBytes) : base(innerStream)
+        {
+            _readAheadBytes = readAheadBytes;
+            _buffers = new ArraySegment<byte>[2];
+            _buffers[0] = new ArraySegment<byte>(new byte[_readAheadBytes], 0, 0);
+            _buffers[1] = new ArraySegment<byte>(new byte[_readAheadBytes], 0, 0);
+            _previousCompletedReadTask = _previousCompletedEnsureBufferedTask = Task.FromResult(readAheadBytes);
+        }
+
+        public override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return Task.FromCanceled<int>(cancellationToken);
+            }
+
+            if (_buffers[0].Count == 0 && _buffers[1].Count == 0)
```

Comment:
The decision to call base.ReadAsync() is based on the buffers both being empty.  Is there any code path that could enter here after _previousCompletedReadTask has been set but before anything is in one of the buffers?  I'm just trying to understand if there are any Asserts we can make to guarantee presumed invariants (like _previousCompletedReadTask not already being set, etc.).